### PR TITLE
PPS: relative alignment

### PR DESCRIPTION
**Pull request by jan-kaspar**
#### PR description:

Following the discussion in the [AlCaDB meeting on 28 Sep](https://indico.cern.ch/event/959393/) this PR contains implementation of the "relative" alignment of PPS. Integration of the "global" part is foreseen soon.

For more details please refer to the README file in "CalibPPS/AlignmentRelative".

#### PR validation:

Various tests are included in the test directory.

### CODE REVIEW

File: DataFormats/CTPPSReco/src/classes_def.xml [hunk 1]
```diff
@@ -40,6 +40,9 @@
   <ioread sourceClass="TotemRPUVPattern" version="[2]" targetClass="TotemRPUVPattern" source="bool fittable" target="fittable_">
   <![CDATA[fittable_ = onfile.fittable;]]>
   </ioread>
+  <ioread sourceClass="TotemRPUVPattern" version="[2]" targetClass="TotemRPUVPattern" source="TotemRPUVPattern::ProjectionType projection" target="projection_">
+  <![CDATA[projection_ = onfile.projection;]]>
+  </ioread>
```
Comment (slava77) [inline DataFormats/CTPPSReco/src/classes_def.xml:45]: this is already in the master.
@jan-kaspar please review the commits and remove this file.
Thank you.

Comment (jan-kaspar) [reply DataFormats/CTPPSReco/src/classes_def.xml:45]: Thanks for pointing this out. I will address this in the next days together with the invalid-state files.

**Comment by jan-kaspar** [comment]
I've just pushed the code with cleaned up history.

For the change in CondFormats/PPSObjects, here's a little [before-after comparison](https://github.com/cms-sw/cmssw/files/5317975/reco_cmp.pdf) showing that no change is generated in PPS reco of Run2 data.

**Comment by christopheralanwest** [comment]
please test

**Comment by christopheralanwest** [comment]
please test

**Comment by silviodonato** [comment]
kind reminder @cms-sw/alca-l2

**Comment by silviodonato** [comment]
@mrodozov could you add the new package?

**Comment by silviodonato** [comment]
@cms-sw/alca-l2

**Comment by yuanchao** [comment]
Hi @jan-kaspar, there are some build warnings found in the static build tests regarding the access of config data. The use of "get" method without token is deprecated, e.g.  https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65a3e0/9962/llvm-analysis/report-d53149.html#EndPath
Could you follow this twiki page and modify accordingly? https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit It would be nice that other newly introduced warnings could also be fixed. For details please refer to here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65a3e0/9962/llvm-analysis/

**Comment by jan-kaspar** [comment]
@yuanchao Sure! I will do it in the next days.

**Comment by jan-kaspar** [comment]
BTW: if I run locally (on LXPLUS) `scram b compile COMPILER=llvm-analyzer` I see less warnings than in the above report. How can I reproduce exactly the report results?

**Comment by yuanchao** [comment]
@jan-kaspar you can find the actual command of the static tests on top of the link as mentioned previously. (https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65a3e0/9962/llvm-analysis/)

**Comment by jan-kaspar** [comment]
> @jan-kaspar you can find the actual command of the static tests on top of the link as mentioned previously. (https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65a3e0/9962/llvm-analysis/)

Thanks @yuanchao! I think that this exactly what I did and still I got less errors. But never mind, I guess I understand the source of the errors and I can fix them.

**Comment by christopheralanwest** [comment]
please test

**Comment by yuanchao** [comment]
Hi @jan-kaspar , there's an issue on the try-catch clause in /CalibPPS/AlignmentRelative/src/LocalTrackFitter.cc (https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65a3e0/10376/llvm-analysis/report-700bcf.html#EndPath). Could you fix it?

**Comment by jan-kaspar** [comment]
> Hi @jan-kaspar , there's an issue on the try-catch clause in /CalibPPS/AlignmentRelative/src/LocalTrackFitter.cc (https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65a3e0/10376/llvm-analysis/report-700bcf.html#EndPath). Could you fix it?

I have actually spent non-trivial time to find a workaround and check whether `ATViA.Invert()` would fail. To me, ROOT doesn't seem to provide a method for this. I've tried using `TMatrixD::Determinant()` but then this needs to be compared to a threshold which would need to be compatible with the one used in the `Invert` method. Looking in the implementation of the latter, in fact, different checks (even on several levels) are used. What would be your suggestion?

File: CalibPPS/AlignmentRelative/src/LocalTrackFitter.cc [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  // evaluate local track parameter estimates (h stands for hat)
+  TMatrixD AT(4, selection.size());
+  AT.Transpose(A);
+  TMatrixD ATViA(4, 4);
+  ATViA = AT * Vi * A;
+  TMatrixD ATViAI(ATViA);
+
+  try {
+    ATViAI = ATViA.Invert();
+  } catch (...) {
```
Comment (christopheralanwest) [inline CalibPPS/AlignmentRelative/src/LocalTrackFitter.cc:111]: Why have you not used the following?
```suggestion
  } catch (cms::Exception& e) {
```
I think that what happens is that the CMSSW framework handles the ROOT errors and throws them as `cms::Exception`. ROOT itself doesn't throw exceptions in these cases. If it is rare to have non-invertible matrices here, I think it's fine to catch the exceptions. But if non-invertible matrices occur regularly, we may not want to have the overhead of throwing an exception.

Comment (jan-kaspar) [reply CalibPPS/AlignmentRelative/src/LocalTrackFitter.cc]: Thanks @christopheralanwest , I've applied your suggestion! Furthermore, I checked with the real-life example in `CalibPPS/AlignmentRelative/test/test_with_data` that the catch block is invoked in less than ~1% calls to `LocalTrackFitter::fitAndRemoveOutliers`. In addition note that this code represents a special calibration that would be run "manually" (i.e. NOT in any central WF) only a couple of times per year.

**Comment by christopheralanwest** [comment]
please test

**Comment by yuanchao** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**